src/Chainweb/Rosetta/Util.hs

```diff
@@ -0,0 +1,686 @@
+    -> Maybe T.Text
+    -- ^ Block Hash
+    -> ExceptT RosettaFailure Handler BlockHeader
+findBlockHeaderInCurrFork cutDb cid someHeight someHash = do
+  latestBlock <- getLatestBlockHeader cutDb cid
+  chainDb <- (cutDb ^? cutDbBlockHeaderDb cid) ?? RosettaInvalidChain
+
+  case (someHeight, someHash) of
+    (Nothing, Nothing) -> pure latestBlock
+    (Just hi, Nothing) -> byHeight chainDb latestBlock hi
+    (Just hi, Just hsh) -> do
+      bh <- byHeight chainDb latestBlock hi
+      bhashExpected <- (blockHashFromText hsh) ?? RosettaUnparsableBlockHash
+      if (_blockHash bh == bhashExpected)
+        then pure bh
+        else throwError RosettaMismatchBlockHashHeight
+    (Nothing, Just hsh) -> do
+      bhash <- (blockHashFromText hsh) ?? RosettaUnparsableBlockHash
+      somebh <- liftIO $ (casLookup chainDb bhash)
+      bh <- somebh ?? RosettaBlockHashNotFound
+      isInCurrFork <- liftIO $ memberOfHeader cutDb cid bhash latestBlock
+      if isInCurrFork
+        then pure bh
+        else throwError RosettaOrphanBlockHash
+  where
+    byHeight db latest hi = do
+      somebh <- liftIO $ seekAncestor db latest (int hi)
+      somebh ?? RosettaInvalidBlockHeight
+
+
+getBlockOutputs
+    :: forall cas
+    . PayloadCasLookup cas
+    => PayloadDb cas
+    -> BlockHeader
+    -> ExceptT RosettaFailure Handler (CoinbaseCommandResult, V.Vector (CommandResult Hash))
+getBlockOutputs payloadDb bh = do
+  someOut <- liftIO $ casLookup payloadDb (_blockPayloadHash bh)
+  outputs <- someOut ?? RosettaPayloadNotFound
+  txsOut <- decodeTxsOut outputs ?? RosettaUnparsableTxOut
+  coinbaseOut <- decodeCoinbaseOut outputs ?? RosettaUnparsableTxOut
+  pure $ (coinbaseOut, txsOut)
+
+  where
+    decodeCoinbaseOut :: PayloadWithOutputs -> Maybe (CommandResult Hash)
+    decodeCoinbaseOut = decodeStrictOrThrow . _coinbaseOutput . _payloadWithOutputsCoinbase
+
+    decodeTxsOut :: PayloadWithOutputs -> Maybe (V.Vector (CommandResult Hash))
+    decodeTxsOut pwo = mapM (decodeStrictOrThrow . _transactionOutputBytes . snd)
+                       (_payloadWithOutputsTransactions pwo)
+
+getTxLogs
+    :: PactExecutionService
+    -> BlockHeader
+    -> ExceptT RosettaFailure Handler (Map TxId [AccountLog])
+getTxLogs cr bh = do
+  someHist <- liftIO $ (_pactBlockTxHistory cr) bh d
+  (BlockTxHistory hist) <- (hush someHist) ?? RosettaPactExceptionThrown
+  case (M.size hist) of
+    0 -> throwError RosettaUnparsableTxLog
+    _ -> pure ()
+  let histParsed = M.mapMaybe (mapM txLogToAccountInfo) hist
+  if (M.size histParsed == M.size hist)
+    then pure histParsed  -- all logs successfully parsed
+    else throwError RosettaUnparsableTxLog
+  where
+    d = (Domain' (UserTables "coin_coin-table"))
+
+
+getHistoricalLookupBalance
+    :: PactExecutionService
+    -> BlockHeader
+    -> T.Text
+    -> ExceptT RosettaFailure Handler Decimal
+getHistoricalLookupBalance cr bh k = do
+  someHist <- liftIO $ (_pactHistoricalLookup cr) bh d key
```

Comment:
Could this be subject to sql injection? Is it okay to use the bare text submitted by the user when its going to be used to form sql statements?

Reply:
Also, in the future you should consider possibly making `_pactHistoricalLookup` return the _last two_ transactions (since your SQL is already a `LIMIT 1` affair). Why? Because we can do deltas with that! 😎

Reply:
@slpopejoy regarding returning the last two transactions, I've made an issue for it here: https://github.com/kadena-io/chainweb-node/issues/1078

@slpopejoy and @emilypi regarding the potential sql injection via account address: I've added the following code to check that the account address is a valid Pact `Name` (since I believe row keys need to be valid `Name` correct?)
https://github.com/kadena-io/chainweb-node/pull/1050/files#diff-04e165a290c33e67270413e0ae8c3d85R515-R519

Reply:
Actually, I need to reconsider this approach. It throws an errors with `"address": "POTENT AF MINER"` on mainnet. This seems to be to strict